### ISSUE TITLE
Consider creating a space-only font

### ISSUE THREAD

**Issue by yamatoiizuka** [yamatoiizuka/palt-typesetting#62]
現状は THIN SPACE (U+2009) で四分アキを実現しているが、
両端揃え (`text-align: justify;`) で行末にきた場合には無視される、
SPACE (U+0020) を使用する方が望ましい。

現状 SPACE を採用していない理由は、Chrome のフォントサイズ 10px 制限に起因している。

**解決策**
通常の半角スペースよりも狭い幅の SPACE (U+0020) グリフのみを持ったフォントデータを作成し、
`.typeset-thin-space` 内で使用する。

**Comment by sharapeco** [comment]
図のようにData URIと `@font-face` の記述込みで1kB前後のwebフォントを作成したのでよければお使いください (出力結果はReleasesから)。

スペースと一応NBSPのグリフを用意してあり、幅は50/1000 emから250/1000 emまで10/1000 em刻みで出力しています。

- [ling-font](https://github.com/sharapeco/ling-font)

<img width="935" alt="image" src="https://github.com/yamatoiizuka/palt-typesetting/assets/1665651/fb701890-3feb-4336-b19b-eb04f5734b6e">

**Comment by yamatoiizuka** [comment]
おわーー早速ありがとうございます！

時間取れるときにゆっくり拝見させていただきます…🙏

**Comment by sharapeco** [comment]
フォントチョットワカルだったのでサクっとできました。

和欧間スペースの幅をパラメータとして受け取って上記のようなスタイルシートを出力できれば便利だなと考えています。WOFF2だと圧縮がかかるので難しいかなと思っていたのですが、出力データを見てみたらグリフデータ以外は圧縮がかからないようなのでJavaScriptで動的に生成できそうです。

もしよろしければその仕組みをお作りします。Minifyして3 kB足らずのコードになると思います。

**Comment by yamatoiizuka** [comment]
@sharapeco こちら遅くなりました！

少し検証していたのですが、ゼロ幅もしくは 1/1000unit の NBSP フォントを、同様の @fontface の形式で書き出すことってできますでしょうか？
```
<span class='typeset-thin-space' style='font-family: ling; letter-spacing: 0.1em'>&nbsp;</span>
```
というように、グリフ幅は `letter-spacing` でコントロールできるといいのかなと。

うまくいけばカーニング周りの issue (#50) も合わせて解決できるのではと考えています。 
お手数おかけいたします！

**Comment by sharapeco** [comment]
はい、可能です。確かに `@font-face` を増やすより letter-spacing でコントロールする方が無駄が少ないですね。

ゼロ幅だとフォントとして正しく扱ってもらえるか分からないので、試して大丈夫だったほうをご案内したいと思います。

ちなみにそのコード例ですが、NBSPを使うと行頭行末に残ってしまうので、分割禁止にしたい箇所以外は単なるスペースでいいのではないでしょうか？

<img width="203" alt="image" src="https://github.com/yamatoiizuka/palt-typesetting/assets/1665651/6e1d8188-3253-4141-b992-e8995dec65a1">

```html
<p><span>Lorem</span><span class="sp"> </span><span>ipsum</span><span class="sp">
</span><span>dolor</span><span class="sp"> </span><span>sit</span><span class="sp">
</span><span>amet</span></p>
```

```css
html {
	font-size: 20px;
	color: #333;
	background: #eee;
}

body {
	margin-inline: auto;
	width: 8rem;
	background: #fff;
}

.sp {
	letter-spacing: 0.5em;
	background: #ff8;
}
```

**Comment by sharapeco** [comment]
普通のフォントでも合成用グリフなど、幅が0になる場合があるので幅ゼロで大丈夫かと思ったのですが、Safariだと `letter-spacing` が効きませんでした。グリフ幅をわざわざゼロにしてあることを考えると、多分これはSafariの動作が正しいんですが、それはさておき。

なので、1/1000 emのデータをご用意しました。パブリックドメインとしますのでご自由にお使いください。

- [ling-1.css](https://gist.github.com/sharapeco/1dd7f4433dad07648e1325e1fa5926ee)

検証した内容はこちらです。

- [ling-0](https://lab.suzume.dev/palt-typesetting/ling-0.html)
- [ling-1](https://lab.suzume.dev/palt-typesetting/ling-1.html)

**Comment by yamatoiizuka** [comment]
検証ページまでありがとうございます！！素晴らしいです！
SPACE と NBSP の使い分けは、ご指摘のとおり少し丁寧に見ないとですね。

通常のスペース（英単語間のセパレータ）に関しては ling フォントを使わず、
もとの欧文書体が持っているスペースが出力される、でよいかなと思っています。

ちなみに Chrome で検証した際、

- ling-0 で `letter-spacing: 0.15em` 
- ling-1 で `letter-spacing: 0.149em`

としたとき、ling-1 のスペースの方が開いて見えるのですが、原因分かりますか…？
もしかして 1000units = 1em ではない、とかでしょうか

**Comment by sharapeco** [comment]
すみません、ゼロ幅で動いたと思ってさらにサイズを削ろうと units per em (UPM) を最小値の16にした方のデータをあげてしまったかもしれません。確認します。

**Comment by sharapeco** [comment]
ついでに誤差を最小にするよう10000 UPMで検証してみました。ファイルサイズは同じです。20px程度だと丸められて1000 UPMとの違いが出ませんでしたが、大きなサイズで効いてくると思います。

ブラウザによってテキストの折り返し位置が違うのはAdobe Jenson部分のサイズの丸め誤差のようですね。[検証してみました](https://lab.suzume.dev/palt-typesetting/ling-1.html)が、どのブラウザでも想定通りの幅になっています。

![macOS Big Sur上でのLing fontの検証](https://github.com/yamatoiizuka/palt-typesetting/assets/1665651/cf732f8c-aea1-409c-a756-a5e90fb5d47e)

**Comment by yamatoiizuka** [comment]
ありがとうございます！！！

バッチリよさそうですね。
こちら使用させていただきます！🙏
